presto-testing-docker/src/main/java/io/prestosql/testing/docker/DockerContainer.java

```diff
@@ -0,0 +1,241 @@
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.MoreCollectors.toOptional;
+import static java.lang.String.format;
+import static java.time.temporal.ChronoUnit.MINUTES;
+import static java.time.temporal.ChronoUnit.SECONDS;
+import static java.util.Objects.requireNonNull;
+
+public final class DockerContainer
+        implements Closeable
+{
+    private static final Logger LOG = Logger.get(DockerContainer.class);
+
+    private static final boolean DEBUG = false;
+
+    private static final String HOST_IP = "127.0.0.1";
+    private final String image;
+    private final Map<String, String> environment;
+    private DockerClient dockerClient;
+    private String containerId;
+
+    private Map<Integer, Integer> hostPorts;
+
+    public DockerContainer(String image, List<Integer> ports, Map<String, String> environment, CheckedConsumer<HostPortProvider> healthCheck)
+    {
+        this.image = requireNonNull(image, "image is null");
+        this.environment = ImmutableMap.copyOf(requireNonNull(environment, "environment is null"));
+        try {
+            startContainer(ports, healthCheck);
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void startContainer(List<Integer> ports, CheckedConsumer<HostPortProvider> healthCheck)
+            throws Exception
+    {
+        dockerClient = DefaultDockerClient.fromEnv().build();
+        if (dockerClient.listImages(DockerClient.ListImagesParam.byName(image)).isEmpty()) {
+            checkState(!image.endsWith("-SNAPSHOT"), "Unavailable snapshot image %s, please build before running tests", image);
+            LOG.info("Pulling image %s...", image);
+            dockerClient.pull(image);
+        }
+        if (DEBUG) {
+            Optional<Container> testingContainer = dockerClient.listContainers().stream()
+                    .filter(container -> container.image().equals(image))
+                    .collect(toOptional());
+            if (testingContainer.isPresent()) {
+                containerId = testingContainer.get().id();
+                LOG.info("Container for %s already exists with id: %s", image, containerId);
+                calculateHostPorts(ports);
+            }
+            else {
+                createContainer(ports);
+            }
+        }
+        else {
+            createContainer(ports);
+        }
+
+        checkState(isContainerUp(), "Container was not started properly");
+
+        LOG.info("Auto-assigned host ports are %s", hostPorts);
+
+        waitForContainer(healthCheck);
```

Comment:
If this fails, will the container be removed?

Reply:
Yes. Thanks for asking. It appeared that when method throws in `abortOn` then `failsafe` continue to retry. I fixed that.